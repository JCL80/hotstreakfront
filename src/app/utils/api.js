import axios from 'axios';

const api = axios.create({
  baseURL: process.env.FRONTEND_URL, // Your Express server
});

export default api;
