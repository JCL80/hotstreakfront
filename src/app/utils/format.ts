export const pct = (n: number | undefined) =>
  isFinite(n!) ? `${n!.toFixed(1)}%` : "—";

export const num = (n: number | undefined, d = 1) =>
  isFinite(n!) ? n!.toFixed(d) : "—";

export const ftPct = (ftm?: number, fta?: number) =>
    fta && fta > 0 ? pct((ftm! / fta) * 100) : "—";
  