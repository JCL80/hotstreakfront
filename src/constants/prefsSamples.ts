import type { HotStreakPrefs } from "@/types/heatIndex";

/* ------------------------------------------------------------------ */
/*  Preset A – Shooting‑centric weights                               */
/*     Heavily rewards efficiency; ignores volume/counting numbers.   */
/* ------------------------------------------------------------------ */

export const SHOOTING_PREFS: HotStreakPrefs = {
    // efficiency tilted
    ts: 2,
    efg: 1.5,
    fgp: 1,
    tpp: 1,
    ftp: 1,

    // rest zero
    pts: 0, reb: 0, ast: 0, oreb: 0, dreb: 0,
    stl: 0, blk: 0, tov: 0, pf: 0, plus_minus: 0, ast_to_tov: 0,

    gamesWindow: 5,
    playoffOnly: false,
};


/* ------------------------------------------------------------------ */
/*  Preset B – Counting‑stat weights                                  */
/*     Rewards box‑score production; efficiency set to zero.          */
/* ------------------------------------------------------------------ */

export const COUNTING_PREFS: HotStreakPrefs = {
    pts: 2,
    reb: 1.5,
    ast: 1.5,
    oreb: 1,
    dreb: 1,
    stl: 1,
    blk: 1,

    // efficiency off
    ts: 0, efg: 0, fgp: 0, tpp: 0, ftp: 0,
    tov: 0, pf: 0, plus_minus: 0, ast_to_tov: 0,

    gamesWindow: 5,
    playoffOnly: false,
};
