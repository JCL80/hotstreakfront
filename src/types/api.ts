/**
 * Box score counting stats for a single game or aggregated span.
 * These come straight from the NBA stats endpoint.
 */
export interface BoxStats {
  /** Points scored */
  pts: number;
  /** Assists */
  ast: number;
  /** Total rebounds */
  reb: number;
  oreb: number;
  dreb: number;

  // Shooting splits
  fgm: number; // Field‑goals made
  fga: number; // Field‑goal attempts
  fgp: number; //
  tpp: number; //
  fg3m: number; // 3‑pt makes
  fg3a: number; // 3‑pt attempts
  ftm: number; // Free‑throws made
  fta: number; // Free‑throw attempts
  ftp: number; // Free-throw percentage

  // Miscellaneous
  stl: number; // Steals
  blk: number; // Blocks
  tov: number; // Turnovers
  plus_minus: number; // Raw +/- from NBA.com
  minutes: number; // Minutes played
  pf: number; // Personal Fouls
}

export interface AdvancedStats extends BoxStats {
  /** True‑shooting percentage */
  ts: number;
  /** Effective FG% */
  efg: number;
  /** Usage rate (needs team context) */
  usg?: number;
  /** Offensive rating (pts/100 poss) */
  ortg?: number;
  /** Defensive rating (pts allowed/100 poss) */
  drtg?: number;

  // Public composite metrics
  bpm?: number;
  vorp?: number;
  ws48?: number;
  epm?: number;
  raptor?: number;
  lebron?: number;
  darko?: number;
}

export interface GameLog {
  PTS: number;
  AST: number;
  REB: number;
  FGM: number;
  FGA: number;
  FG3M: number;
  FG3A: number;
  FTM: number;
  FTA: number;
  STL: number;
  BLK: number;
  TOV: number;
  PLUS_MINUS?: number;
  MIN: number; // minutes string like "34:12" or numeric string
  GAME_DATE?: string;
  MATCHUP?: string;
  WL?: string;
  FG_PCT?: number;
  FG3_PCT?: number;
  FT_PCT?: number;
  OREB?: number; // <-- NEW
  DREB?: number; // <-- NEW
  PF?: number;
}

/**
 * Generic result set coming from your FastAPI backend.
 */
export interface ResultSet {
  name: string;
  headers: string[];
  rowSet: unknown[][];
}

/**
 * Shape of the JSON from GET /player/{id}
 */
export interface PlayerResponse {
  resultSets: ResultSet[];
  image_url: string;
  team_id: number; // <-- Add this
  team_logo_url: string; // <-- Add this
}

/**
 * Structured player bio.
 */
export interface PlayerBio {
  id: number;
  first_name: string;
  last_name: string;
  team: { id: number; full_name: string; logo_url: string }; // Updated
  position: string;
  height: string;
  weight: string;
  jersey_number: number;
  college: string;
  country: string;
  image_url: string;
}

/**
 * Shape of the JSON from GET /player/{id}/games
 */
export interface GamesResponse {
  resultSets: ResultSet[];
}
