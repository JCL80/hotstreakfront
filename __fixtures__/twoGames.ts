// __tests__/fixtures/twoGames.ts
import { GameLog } from "@/types/api";

export const g1: GameLog = {
  PTS: 20, AST: 4, REB: 5,
  FGM: 8,  FGA: 15,
  FG3M: 2, FG3A: 6,
  FTM: 2,  FTA: 2,
  STL: 1,  BLK: 0, TOV: 2,
  PLUS_MINUS: 5,
  MIN: 30,
  GAME_DATE: "2024-01-01",
  MATCHUP: "TEAM vs. OPP",
  WL: "W",
  FG_PCT: 53.3,
  FG3_PCT: 33.3,
  FT_PCT: 100.0,
  OREB: 2,
  DREB: 3,
  PF: 2,
  PLAYOFF: false
};

export const g2: GameLog = {
  ...g1,
  PTS: 30, // bump a few fields to be different
  FGM: 12,
  PLUS_MINUS: -3,
  MIN: 28,
  GAME_DATE: "2024-01-02",
  MATCHUP: "TEAM @ OPP",
  WL: "L",
  FG_PCT: 60.0,
  FG3_PCT: 40.0,
  FT_PCT: 80.0,
  OREB: 3,
  DREB: 4,
  PF: 3
};
