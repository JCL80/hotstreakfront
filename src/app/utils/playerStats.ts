/* eslint-disable @typescript-eslint/no-unused-vars */

import { AdvancedStats } from "@/types/api";
import { HeatWeights, HotStreakPrefs } from "@/types/heatIndex";
import { STAT_KEYS , STAT_MIN_MAX , NEGATIVE_STATS} from "../../constants/statKeys";


export function calculateHeatIndex(
  stats: AdvancedStats,
  weights: HeatWeights
): number {
  // If no weights are provided, return 0
  if (!weights || Object.keys(weights).length === 0) {
    return 0
  }

  // Check if all stats are at minimum
  const isAllMin = Object.entries(stats).every(([key, value]) => {
    if (!(key in STAT_MIN_MAX)) return true // Skip stats not in STAT_MIN_MAX
    const [min, _] = STAT_MIN_MAX[key as keyof typeof STAT_MIN_MAX]
    return value === min || value === 0
  })

  if (isAllMin) {
    return 0
  }

  // Check if all stats are at or near maximum
  const isAllMax = Object.entries(stats).every(([key, value]) => {
    if (!(key in STAT_MIN_MAX)) return true // Skip stats not in STAT_MIN_MAX
    const [_, max] = STAT_MIN_MAX[key as keyof typeof STAT_MIN_MAX]
    
    // For percentage stats, check if value is at or above max
    if (['ts', 'efg', 'fgp', 'tpp', 'ftp'].includes(key)) {
      return value >= max * 0.9 // Allow for small variations
    }
    
    // For negative stats, 0 is best
    if (NEGATIVE_STATS.has(key as keyof typeof STAT_MIN_MAX)) {
      return value <= max * 0.1 // Allow for small variations
    }
    
    // For other stats, check if value is at or above max
    return value >= max * 0.9 // Allow for small variations
  })

  if (isAllMax) {
    return 100
  }

  let weightedSum = 0
  let totalWeight = 0

  for (const key of STAT_KEYS) {
    const rawValue = stats[key] ?? 0
    const [min, max] = STAT_MIN_MAX[key]
    
    // Calculate how good the stat is (0 to 1)
    let normalizedValue
    if (NEGATIVE_STATS.has(key)) {
      // For negative stats (turnovers, fouls), 0 is best
      normalizedValue = 1 - Math.min(rawValue / max, 1)
    } else if (['ts', 'efg', 'fgp', 'tpp', 'ftp'].includes(key)) {
      // For percentage stats, normalize based on min/max
      normalizedValue = Math.min((rawValue - min) / (max - min), 1)
    } else {
      // For other stats, normalize based on min/max
      normalizedValue = Math.min((rawValue - min) / (max - min), 1)
    }

    // Keep value between 0 and 1
    normalizedValue = Math.min(Math.max(normalizedValue, 0), 1)

    // Add weighted contribution
    const weight = weights[key] ?? 0
    weightedSum += normalizedValue * weight
    totalWeight += weight
  }

  // If all weights are 0, return 0
  if (totalWeight === 0) {
    return 0
  }

  // Calculate final score (0 to 100)
  const average = weightedSum / totalWeight
  return Math.round(average * 100)
}

export const getHotColdStreak = async (
  prefs: HotStreakPrefs,
  seasonAverages: AdvancedStats | null,
  lastNGamesAvg: AdvancedStats | null
): Promise<{
  status: string
  difference: number
}> => {
  // Handle missing data first
  if (!seasonAverages || !lastNGamesAvg || !prefs || Object.keys(seasonAverages).length === 0 || Object.keys(lastNGamesAvg).length === 0) {
    return {
      status: 'No Data',
      difference: 0,
    }
  }

  try {
    const { gamesWindow, ...weights } = prefs
    
    // Check if weights are valid
    if (Object.keys(weights).length === 0) {
      return {
        status: 'Error',
        difference: 0,
      }
    }
    
    // Calculate heat indexes
    const seasonHeat = calculateHeatIndex(seasonAverages, weights)
    const recentHeat = calculateHeatIndex(lastNGamesAvg, weights)
    const difference = Math.round((recentHeat - seasonHeat) * 10) / 10

    // Determine hot/cold status
    let status
    if (difference > 5) status = 'Hot'
    else if (difference < -5) status = 'Cold'
    else status = 'Neutral'

    return { status, difference }
  } catch (err) {
    console.error('Hot/cold streak error:', err)
    return {
      status: 'Error',
      difference: 0,
    }
  }
}

