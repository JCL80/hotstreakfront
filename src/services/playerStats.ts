import { AdvancedStats, BoxStats, GameLog } from "@/types/api";

export const trueShooting = (pts: number, fgm: number, fta: number) =>
  pts / (2 * (fgm + 0.44 * fta));

export const effectiveFg = (fgm: number, fg3m: number, fga: number) =>
  (fgm + 0.5 * fg3m) / fga;

/** Convert NBA "MIN" field to decimal minutes (e.g. "34:12" → 34.2). */
export const parseMinutes = (raw: unknown): number => {
  if (raw === null || raw === undefined) return 0;

  /* ---------- 1. Input is already a number ---------- */
  // Reject NaN and negative values; keep 0-48+ as-is.
  if (typeof raw === 'number') {
    return isFinite(raw) && raw >= 0 ? raw : 0;
  }

  const str = String(raw).trim();

  /* ---------- 2. Plain integer string ---------- */
  if (/^\d+$/.test(str)) {
    return Number(str);              // "23" => 23
  }

  /* ---------- 3. Colon formats ---------- */
  // Accept mm:ss  (e.g. 12:45)  or  h:mm:ss  (e.g. 1:02:30)
  if (/^\d+:\d{1,2}(:\d{2})?$/.test(str)) {
    const parts = str.split(':').map(Number);      // ["34","12"] → [34,12]

    let hours = 0, minutes = 0, seconds = 0;
    if (parts.length === 3) {          // h:mm:ss
      [hours, minutes, seconds] = parts;
    } else {                           // mm:ss
      [minutes, seconds] = parts;
    }

    return hours * 60 + minutes + seconds / 60;
  }

  /* ---------- 4. Everything else ---------- */
  // "DNP", "—", empty, garbage ⇒ 0
  return 0;
};



export const normalizeGameLog = (raw: Record<string, unknown>): GameLog => {
  return {
    PTS: Number(raw["PTS"]) || 0,
    AST: Number(raw["AST"]) || 0,
    REB: Number(raw["REB"]) || 0,
    FGM: Number(raw["FGM"]) || 0,
    FGA: Number(raw["FGA"]) || 0,
    FG3M: Number(raw["FG3M"]) || 0,
    FG3A: Number(raw["FG3A"]) || 0,
    FTM: Number(raw["FTM"]) || 0,
    FTA: Number(raw["FTA"]) || 0,
    STL: Number(raw["STL"]) || 0,
    BLK: Number(raw["BLK"]) || 0,
    TOV: Number(raw["TOV"]) || 0,
    PLUS_MINUS:
      raw["PLUS_MINUS"] !== undefined ? Number(raw["PLUS_MINUS"]) : undefined,
    MIN: parseMinutes(raw["MIN"]),
    GAME_DATE : String(raw["GAME_DATE"] ?? ''),
    MATCHUP   : String(raw["MATCHUP"]   ?? ''),
    WL        : String(raw["WL"]        ?? ''),
    FG_PCT    : Number(raw["FG_PCT"]    ?? 0),
    FG3_PCT   : Number(raw["FG3_PCT"]   ?? 0),
    FT_PCT    : Number(raw["FT_PCT"]    ?? 0),
  };
};

export const getAverages = (games: GameLog[]): AdvancedStats => {
  const totals = {
    pts: 0,
    ast: 0,
    reb: 0,
    oreb: 0,
    dreb: 0,
    fgm: 0,
    fga: 0,
    fg3m: 0,
    fg3a: 0,
    ftm: 0,
    fta: 0,
    stl: 0,
    blk: 0,
    tov: 0,
    pf: 0,
    plus_minus: 0,
    minutes: 0,
  };

  for (const g of games) {
    totals.pts += g.PTS;
    totals.ast += g.AST;
    totals.reb += g.REB;
    totals.oreb += g.OREB ?? 0;
    totals.dreb += g.DREB ?? 0;
    totals.fgm += g.FGM;
    totals.fga += g.FGA;
    totals.fg3m += g.FG3M;
    totals.fg3a += g.FG3A;
    totals.ftm += g.FTM;
    totals.fta += g.FTA;
    totals.stl += g.STL;
    totals.blk += g.BLK;
    totals.tov += g.TOV;
    totals.pf += g.PF ?? 0;
    totals.plus_minus += g.PLUS_MINUS ?? 0;
    totals.minutes += g.MIN;
  }

  const c = games.length || 1;

  const box: BoxStats = {
    pts: totals.pts / c,
    ast: totals.ast / c,
    reb: totals.reb / c,
    oreb: totals.oreb / c,
    dreb: totals.dreb / c,
    fgm: totals.fgm / c,
    fga: totals.fga / c,
    fgp: totals.fga ? (totals.fgm / totals.fga) * 100 : 0,
    fg3m: totals.fg3m / c,
    fg3a: totals.fg3a / c,
    tpp: totals.fg3a ? (totals.fg3m / totals.fg3a) * 100 : 0,
    ftm: totals.ftm / c,
    fta: totals.fta / c,
    ftp: totals.fta ? (totals.ftm / totals.fta) * 100 : 0,
    stl: totals.stl / c,
    blk: totals.blk / c,
    tov: totals.tov / c,
    pf: totals.pf / c,
    plus_minus: totals.plus_minus / c,
    minutes: totals.minutes / c,
  };

  const tsDen = box.fgm + 0.44 * box.fta;

  const finalStats: AdvancedStats = {
    ...box,
    ts: tsDen ? trueShooting(box.pts, box.fgm, box.fta) * 100 : 0, // % True Shooting
    efg: box.fga ? effectiveFg(box.fgm, box.fg3m, box.fga) * 100 : 0, // % Effective FG
  };

  return finalStats;
};

