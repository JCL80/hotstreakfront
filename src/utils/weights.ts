import { HotStreakPrefs } from "@/types/heatIndex";

export const stripWeights = (prefs: HotStreakPrefs) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { gamesWindow, playoffOnly, ...weights } = prefs;
  return weights;
};
