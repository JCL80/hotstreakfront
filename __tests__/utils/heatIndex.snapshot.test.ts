import { calculateHeatIndex } from '@/app/utils/playerStats';
import { stripWeights } from '@/utils/weights';     // helper we built earlier
import { AdvancedStats }      from '@/types/api';
import { jokicSeasonAvg } from '../../__fixtures__/realisticGames';
import { SHOOTING_PREFS, COUNTING_PREFS } from '@/constants/prefsSamples';

/**
 * Two meaningful prefs presets
 * 1. Shooting‑only   → emphasises TS, eFG, FG%, 3P%, FT%
 * 2. Counting stats  → emphasises PTS, REB, AST, STL, BLK
 */

describe('calculateHeatIndex snapshots', () => {
  const shootingScore = calculateHeatIndex(
    jokicSeasonAvg as AdvancedStats,
    stripWeights(SHOOTING_PREFS),
  );

  const countingScore = calculateHeatIndex(
    jokicSeasonAvg as AdvancedStats,
    stripWeights(COUNTING_PREFS),
  );

  it('matches snapshot – shooting‑only weighting', () => {
    expect(shootingScore).toMatchSnapshot();
  });

  it('matches snapshot – counting‑stats weighting', () => {
    expect(countingScore).toMatchSnapshot();
  });
});
