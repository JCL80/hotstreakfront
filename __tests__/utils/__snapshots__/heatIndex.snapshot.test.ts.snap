// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateHeatIndex snapshots matches snapshot – counting‑stats weighting 1`] = `48`;

exports[`calculateHeatIndex snapshots matches snapshot – shooting‑only weighting 1`] = `87`;
